pkg/services/ngalert/api/tooling/definitions/alertmanager_validation.go

```diff
@@ -0,0 +1,65 @@
+package definitions
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/prometheus/common/model"
+)
+
+// Validate normalizes a Route r, and returns errors if r is invalid.
+func (r *Route) Validate() error {
+	for _, l := range r.GroupByStr {
+		if l == "..." {
+			r.GroupByAll = true
+		} else {
+			r.GroupBy = append(r.GroupBy, model.LabelName(l))
+		}
+	}
+
+	if len(r.GroupBy) > 0 && r.GroupByAll {
+		return fmt.Errorf("cannot have wildcard group_by (`...`) and other other labels at the same time")
+	}
+
+	groupBy := map[model.LabelName]struct{}{}
+
+	for _, ln := range r.GroupBy {
+		if _, ok := groupBy[ln]; ok {
+			return fmt.Errorf("duplicated label %q in group_by", ln)
+		}
+		groupBy[ln] = struct{}{}
+	}
+
+	if r.GroupInterval != nil && time.Duration(*r.GroupInterval) == time.Duration(0) {
+		return fmt.Errorf("group_interval cannot be zero")
+	}
+	if r.RepeatInterval != nil && time.Duration(*r.RepeatInterval) == time.Duration(0) {
+		return fmt.Errorf("repeat_interval cannot be zero")
+	}
+
+	// Routes are a self-referential structure.
+	if r.Routes != nil {
+		for _, child := range r.Routes {
+			err := child.Validate()
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// ValidateRoot normalizes a Route r, and returns errors if r is an invalid root route. Root routes must satisfy a few additional conditions.
+func (r *Route) ValidateRoot() error {
```

Comment:
Rather than having two validate calls, I think a better architecture would be to have the root route be a different type altogether, that just excludes the appropriate fields. However, that is a pretty significant change that doesn't seem suited for the goal of this PR. I still think this is a step in the right direction.

Reply:
are both methods `Validate` and `ValidateRoot` used outside the package? If no, probably make former private?

Reply:
Good idea.